airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py

```diff
@@ -84,3 +82,109 @@ def access_token(self) -> str:
         self._access_token = value
+
+
+class SingleUseRefreshTokenOauth2Authenticator(Oauth2Authenticator):
+    """
+    Authenticator that should be used for API implementing single use refresh tokens:
+    when refreshing access token some API returns a new refresh token that needs to used in the next refresh flow.
+    This authenticator updates the configuration with new refresh token by emitting Airbyte control message from an observed mutation.
+    This authenticator expects a connector config with a"credentials" field with the following nested fields: client_id, client_secret, refresh_token.
+    This behavior can be changed by overriding getters or changing the default "credentials_configuration_field_name" value.
+    """
+
+    def __init__(
+        self,
+        connector_config: Mapping[str, Any],
+        token_refresh_endpoint: str,
+        scopes: List[str] = None,
+        token_expiry_date: pendulum.DateTime = None,
+        access_token_name: str = "access_token",
+        expires_in_name: str = "expires_in",
+        refresh_token_name: str = "refresh_token",
+        refresh_request_body: Mapping[str, Any] = None,
+        grant_type: str = "refresh_token",
+        credentials_configuration_field_name: str = "credentials",
+    ):
+        self.credentials_configuration_field_name = credentials_configuration_field_name
+        self._refresh_token_name = refresh_token_name
+        self._connector_config = observe_connector_config(connector_config)
+        self._validate_connector_config()
+        super().__init__(
+            token_refresh_endpoint,
+            self.get_client_id(),
+            self.get_client_secret(),
+            self.get_refresh_token(),
+            scopes,
+            token_expiry_date,
+            access_token_name,
+            expires_in_name,
+            refresh_request_body,
+            grant_type,
+        )
+
+    def _validate_connector_config(self):
+        """Validates the defined getters for configuration values are returning values.
+
+        Raises:
+            ValueError: Raised if the defined getters are not returning a value.
+        """
+        for field_name, getter in [
+            ("client_id", self.get_client_id),
+            ("client_secret", self.get_client_secret),
+            (self.get_refresh_token_name(), self.get_refresh_token),
+        ]:
+            try:
+                assert getter()
+            except (AssertionError, KeyError):
+                raise ValueError(
+                    f"This authenticator expects a {field_name} field under the {self.credentials_configuration_field_name} field. Please override this class getters or change your configuration structure."
+                )
+
+    def get_refresh_token_name(self) -> str:
+        return self._refresh_token_name
+
+    def _get_config_credentials_field(self, field_name):
+        return self._connector_config[self.credentials_configuration_field_name][field_name]
```

Comment:
we should use `dpath` to allow easily removing any assumptions about where in the code the refresh/access tokens live

Reply:
Thanks, that was definitely what I was missing, I was not aware of the existence of this lib 😄